[subclasses] Do not fakeTensor const prop subclass args

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134855

The issue:

Const propagation checks only if arguments do not have FakeTensor. If argument is Subclass, it will pass this condition.

As a result Const Propogation execution happens without FakeTensorMode and having tensor factories inside Subclass.__torch_dispatch__ results that this Tensor is not Fakified.

Solution:

If we have subclasses arguments, do not count that const propagation is doable

IvanKobzarev committed:
[subclasses] Do not fakeTensor const prop subclass args

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[subclasses] Do not fakeTensor const prop subclass args"




[ghstack-poisoned]

zou3519 commented on test/functorch/test_aotdispatch.py:
```diff
@@ -6010,6 +6010,18 @@ def forward(self, x):
+
+        ref_out = fn(inp)
+        out = torch.compile(fn, backend="aot_eager", fullgraph=True)(inp)
+        # self.assertEqual(ref_out, out)
```
what's up here?

IvanKobzarev replied:
Yeah, I commented it, forgot to remove. 
Assert fails on some op that is not implemented in my Testing Subclass.

IvanKobzarev replied:
Resolved, changed impl fn to * ones only for float32

zou3519 commented on torch/testing/_internal/test_subclasses.py:
```diff
@@ -0,0 +1,53 @@
+import torch.utils._pytree as pytree
+
+
+class SubclassWithTensorFactory(torch.Tensor):
```
this file should go in https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_subclass.py

zou3519 reviewed (dismissed):
avoiding constant prop on subclasses seems reasonable to me. Left some minor comments, please read

zou3519 reviewed (commented):
Sorry, I changed my mind. It's weird to me that Tensor subclasses are showing up in FakeTensor's `__torch_dispatch__`, I would expect that it never sees subclasses

zou3519 reviewed (approved):
JK, this is FakeTensorMode

IvanKobzarev committed:
Update on "[subclasses] Do not fakeTensor const prop subclass args"


The issue:

Const propagation checks only if arguments do not have FakeTensor. If argument is Subclass, it will pass this condition.

As a result Const Propogation execution happens without FakeTensorMode and having tensor factories inside Subclass.__torch_dispatch__ results that this Tensor is not Fakified.

Solution:

If we have subclasses arguments, do not count that const propagation is doable


[ghstack-poisoned]

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1736,6 +1737,7 @@ def _dispatch_impl(
             should_allow_numbers_as_tensors(func)
             and not has_symbolic_sizes
             and not flat_arg_fake_tensors
+            and not has_subclasses
```
Yeah, I also agree this is a bandaid fix and at least a comment saying how to do it properly would be good.

IvanKobzarev commented:
@pytorchbot merge
